cpp/src/slice2java/Gen.cpp

```diff
@@ -3703,17 +3703,13 @@ void
 Slice::Gen::HelperVisitor::visitSequence(const SequencePtr& p)
 {
     BuiltinPtr builtin = dynamic_pointer_cast<Builtin>(p->type());
-    if (!hasTypeMetadata(p) && builtin && builtin->kind() <= Builtin::KindString)
+    bool mappedToCustomType = p->hasMetadata("java:type");
+    if (builtin && builtin->kind() < Builtin::KindObject && !mappedToCustomType)
     {
-        return; // No helpers for sequence of primitive types
-    }
-    else if (hasTypeMetadata(p) && !p->hasMetadata("java:type"))
-    {
-        return; // No helpers for custom metadata other than java:type
```

Comment:
Yes, that's correct.
For completeness (and future readers) though:

The first `if` check is completely unchanged. And the 2nd one is redundant with these new guarantees.
You can 'expand' the 'hasTypeMetadata':
```
(hasTypeMetadata(p) && !p->hasMetadata("java:type"))
    ==>
((p->hasMetadata("java:buffer") || p->hasMetadata("java:serializable")) && !p->hasMetadata("java:type"))
```
(hasTypeMetadata also checks for `java:type`, but we can ignore it because of the `&& !p->hasMetadata("java:type")`)

And we already have a variable for saying that `java:type` doesn't appear, so:
```
((p->hasMetadata("java:buffer") || p->hasMetadata("java:serializable")) && !p->hasMetadata("java:type"))
    ==>
((p->hasMetadata("java:buffer") || p->hasMetadata("java:serializable")) && !mappedToCustomType)
```
This looks very similar to the first if check of:
```
if (builtin && builtin->kind() < Builtin::KindObject && !mappedToCustomType)
```

And it is in fact, completely redundant. Because, like you point out, `java:buffer` and `java:serializable` can only appear on sequences of builtin types. so if either of these are present, _we would of already gotten into the first if block_.